### PR TITLE
Issue #36 and #37 - Field Help and Auto Generate ArgumentParser

### PR DESCRIPTION
**Pull request by ameily**
This PR adds `Field.help` /  `Field.short_help` (#36) and adds `Schema.generate_argparse_parser()` (#37).

I've held off on adding unit tests around the `generate_argparse_parser()` until I get feedback on the design.

### CODE REVIEW

**Comment by ameily** [comment]
@Rudedog9d can you take a look at this to see if this matches your needs? Right now the `Schema.generate_argparse_parser()`:

- Generates `--long-opt-names`
- Only generates argument for string, int, float, and bool fields
- Bool fields have two arguments generated `--long-opt` (`True`) and `--no-long-opt` (`False`)
- The `Field.short_help` is included in each argument.
- No validation is performed on argument values, since it's the config that will actually perform the validation (via `cmdline_args_override()`)

File: cincoconfig/config.py [hunk 1]
```diff
@@ -309,6 +309,33 @@ def wrapper(meth: ConfigInstanceMethod) -> ConfigInstanceMethod:
...
+        parser = ArgumentParser(**kwargs)
+        for name, _, field in self.get_all_fields():
+            if not isinstance(field, Field):
+                continue
+
+            arg = '--' + name.replace('.', '-').replace('_', '-').lower()
+            metavar = name.replace('.', '_').upper()
+            if field.storage_type in (str, float, int):
+                parser.add_argument(arg, action='store', dest=name, help=field.short_help,
+                                    metavar=metavar)
```
Comment (Rudedog9d) [inline cincoconfig/config.py:331]: Would you consider adding support for environment variables to argparse? (particularly useful for secrets in pipelines or containerized environments)

```suggestion
            long_arg = '--' + name.replace('.', '-').replace('_', '-').lower()
            env_var_name = name.replace('.', '_').upper()
            metavar = name.replace('.', '_').upper()
            if field.storage_type in (str, float, int):
                parser.add_argument(long_arg, action='store', dest=name, help=field.short_help,
                                    metavar=metavar, default=os.environ.get(env_var_name))
```

Comment (ameily) [reply cincoconfig/config.py:331]: Hmmm, is there a reason you'd want to set the default in the `ArgumentParser` and not the Field itself?

```python
schema.password = FIeld(default=os.environ.get('DATABASE_PASSWORD'))
```

If an argument isn't specified it'll use the `Field.default`.

**Comment by Rudedog9d** [comment]
@ameily This is awesome!! I love the ability to dump the help aswell. I tested this with a nested config, which worked exactly as I'd expect (`alpaca.data_url` became `--alpaca-data-url`.

This unfortunately doesn't solve my original motivation for issue #36, but it's still two great features. Originally, was to add the `help` value as comments to my yaml config. I looked into this a bit myself, and it's possible ([ruamel.yaml](https://pypi.org/project/ruamel.yaml) supports [reading and writing comments](https://stackoverflow.com/questions/46786876/edit-existing-yaml-file-but-keeping-original-comments), for example) - but it sounds like a huge pain, and probably not worth the effort.

**Comment by ameily** [comment]
Ahhh, I got all excited about auto generating a command line parser that I forgot you wanted to write out the help as comments. Yes, that's something I can look into.

**Comment by Rudedog9d** [comment]
@ameily it would also be awesome to preserve existing comments in the config (if an option is commented out, or someone adds a custom comment about something) - so maybe loading/saving only preserves comments, and the help is added if a config is generated the first time? Or maybe a comment is only added before an option if there isn't an existing comment?

File: cincoconfig/config.py (hunk 1 above)
Comment (Rudedog9d) [reply cincoconfig/config.py:331]: Yes, the default value in the config schema should be the application default, ex:

```py
Schema.log_level_console = LogLevelField(default='info')
```

You could do `default=os.environ.get(metavar, 'info')` I suppose, but I thought it would be nice to have the environment variable code auto-generated, so the whole config isn't filled with `os.environ.get('CAP_STR')` entries.

Setting it in argparse isn't actually setting a default, since `os.environ.get()` will return NONE if the variable isn't defined, and therefore the config item won't be overridden in `cmdline_args_override`

**Comment by ameily** [comment]
@Rudedog9d Comments are hard apparently. It is shocking that:

- YAML is wayyy more complex than I was aware.
-  No library exists to keep, edit/replace, and write comments

I would love to know how something like VSCode does it because when I modify my setting in the UI, the config file isn't reordered and my comments aren't removed.

This may be a little ambitious, but my current thought is to create a restricted YAML format that only supports list, str, int, float, bool, and dict. The parser would preserve field order and comments. The only issue I foresee is that if the config file is written or modified and puts in a YAML compatible feature then the ryaml parser would fail.

**Comment by vix597** [comment]
I would just keep the comments/help messages in the `config.py` file itself and say we don't write comments or help to any config files yet. They're still useful to have even if we never write them to config files.

My other thought: Can you do something like read/write the comments yourself but remove them before it hits the real parser. Like for example, yaml comments seem simple, they just begin with `#`. So you could:

### On Save

1. Use the yaml parser to write the file to disk
1. Open the file, read it back in, go line-by-line and place the comments where they belong

### On Load

1. Read the file in and remove the comments you placed
1. Then pass it to the proper yaml parser

I realize after typing that out that it'll probably be a lot harder to do what I've suggested on save than it would be to remove comments on load so maybe this is overkill.

Another thought, you could write a kind-of "Config-help" file right next to the config called `<config-filename>.cfg.help` or something and that would contain all the config field names and help messages formatted as like markdown or something. Then we could benefit from the help messages no matter what the config format is.

TLDR: 
1. Don't write comments at all
1. Write them but sneaky strip them out before and after the parser
1. Write a whole other help file.

**Comment by Rudedog9d** [comment]
@ameily comments sound like a huge pain, lets drop the idea for now. Maybe it's worth opening a separate issue and pondering the idea, but for my usecase, it's not worth the work.

Great work on the argparse stuff, this is an awesome feature!

**Comment by ameily** [comment]
I'm going to make a separate issue for our discussion. I think the solution would be to make our own cincoconfig file format that is the most feature rich:

- Preserve comments
- Write new comments (field documentation)
- Better parse error messages (line / column / example)
- Simplified YAML
